provisioner/config.go

```diff
@@ -197,18 +243,127 @@ func (c *VolumeConfig) GetNFServerGraceTime() (int, error) {
+//                      GID: <group-ID>
+//                      mode: "g+s"
+// -----------------------------------------------------
 func (c *VolumeConfig) GetFSGroupID() (*int64, error) {
-	fsGIDStr := c.getValue(FSGroupID)
-	if len(strings.TrimSpace(fsGIDStr)) == 0 {
+	fsGroupIDStr := c.getValue(FSGroupID)
+	if len(strings.TrimSpace(fsGroupIDStr)) == 0 {
 		return nil, nil
 	}
-	fsGIDInt, err := strconv.ParseInt(fsGIDStr, 10, 64)
+	fsGIDInt, err := strconv.ParseInt(fsGroupIDStr, 10, 64)
 	if err != nil {
 		return nil, err
 	}
+
+	klog.Infof("The %s option key '%s' is being deprecated"+
+		" and will be removed in future releases."+
+		"\nYou may use the %s option key in the "+
+		"backend volume PersistentVolumeClaim's %s "+
+		"annotation key to achieve the same result."+
+		"\nSample config:\n"+
+		"\t\t"+"- name: FilePermissions\n"+
+		"\t\t"+"  data:\n"+
+		"\t\t"+"    %s: \"%s\"\n"+
+		"\t\t"+"    %s: \"g+s\"\n",
+		string(mconfig.CASConfigKey), FSGroupID, FilePermissions,
+		string(mconfig.CASConfigKey), FsGID, fsGroupIDStr, FsMode,
+	)
 	return &fsGIDInt, nil
 }
 
+// GetFsGID fetches the group owner's ID from
+// PVC annotation, if specified
+func (c *VolumeConfig) GetFsGID() (string, error) {
+	fsGIDStr := strings.TrimSpace(c.getData(FilePermissions, FsGID))
+
+	// TODO: remove this block when FSGID is deprecated
+	// Replace this block with the commented code below
+	fsGroupIDStr, _ := c.GetFSGroupID()
+
+	existsFsGIDStr := (len(fsGIDStr) > 0)
+	existsFsGroupIDStr := (fsGroupIDStr != nil)
+
+	// Checking if FSGID and FilePermissions (GID) are being used together
+	if existsFsGIDStr && existsFsGroupIDStr {
+		return "", errors.Errorf("both '%s' and %s data "+
+			"key '%s' cannot be used together",
+			FSGroupID, FilePermissions, FsGID,
+		)
+	}
+
+	if existsFsGroupIDStr {
+		return "", nil
+	} else if existsFsGIDStr {
+		return fsGIDStr, nil
+	}
+
+	return "", nil
+
+	/*
+		if len(fsGIDStr) == 0 {
+			return ""
+		}
+
+		return fsGIDStr
+	*/
+}
+
+// GetFsGID fetches the user owner's ID from
+// PVC annotation, if specified
+func (c *VolumeConfig) GetFsUID() string {
+	fsUIDStr := strings.TrimSpace(c.getData(FilePermissions, FsUID))
+	if len(fsUIDStr) == 0 {
+		return ""
+	}
+
+	return fsUIDStr
+}
+
+// GetFsMode fetches the file mode from
+// PVC annotation, if specified
+func (c *VolumeConfig) GetFsMode() (string, error) {
+	modeStr := strings.TrimSpace(c.getData(FilePermissions, FsMode))
+
+	// TODO: remove this block when FSGID is deprecated
+	// Replace this block with the commented code below
+	fsGroupIDStr, _ := c.GetFSGroupID()
+
+	existsFsModeStr := (len(modeStr) > 0)
+	existsFsGroupIDStr := (fsGroupIDStr != nil)
+
+	// Checking if FSGID and FilePermissions (mode) are being used together
+	if existsFsModeStr && existsFsGroupIDStr {
+		return "", errors.Errorf("both '%s' and %s data "+
+			"key '%s' cannot be used together",
+			FSGroupID, FilePermissions, FsMode,
+		)
+	}
+
+	if existsFsGroupIDStr {
+		return "", nil
```

Comment:
We are using FSGID.

If we have reached line 352 in the function, then we know that both FSGID and FilePermissions-GID both don't exist. So we need not fail and return error.

If FSGID exists, the the GetFSGroupID() function will be used to fetch the FSGID value. 

In line 352, we are returning `"",nil` if FSGID exists because the GetFsGID() function isn't going to be used to fetch the FSGID value. That is being done by GetFSGroupID().

GetFsGID() is only concerned with 2 things:
1. Checking if FSGID and FilePermissions-GID is used together and failing.
2. Fetching FilePermissions-GID in case the above is not true.

Reply:
> If (1) is true and FSGID exists, then GetFsGID() simply returns an empty string.

So we are not honoring user configuration(if FSGID is set on StorageClass), instead of silently ignoring configuration we can log it as a warning.

Reply:
Yes we are. The FSGID value will be fetched using a different method --> GetFSGroupID()

Reply:
Ok, got your point we have existing code which uses FSGID via `GetFSGroupID()` at [271](https://github.com/openebs/dynamic-nfs-provisioner/blob/347ab2f6c10d8faeb209743bd0dfe6d65d76a52f/provisioner/helper_kernel_nfs_server.go#L271).